### ISSUE TITLE
Error in extractTemplatesfromSnippets

### ISSUE THREAD

**Issue by smasri09** [MouseLand/Kilosort#335]
Hello, I am not yet sure if this error is my fault of the software, but I thought I'd post here while I continue to debug it.  I'm processing 64ch data recorded from a whitematter system, read into a matlab array and saved an int16 file from matlab.  Using the latest version of kilosort.

I'm getting this error:

Error using gpuArray/subsref
Subscript indices must either be real positive integers or logicals.

Error in extractTemplatesfromSnippets (line 56)
wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));

Error in datashift2 (line 4)
[wTEMP, wPCA]    = extractTemplatesfromSnippets(rez, NrankPC);

Error in ksGUI/runPreproc (line 726)
                obj.rez = datashift2(obj.rez, 1);

Error in ksGUI>@(~,~)obj.runPreproc() (line 331)
                'Callback', @(~,~)obj.runPreproc());
 
Error while evaluating UIControl Callback.

On debugging, I'm noticing that 1) dd is a large array of zeros, and 2) this line 

% discard empty samples
dd = dd(:, 1:k);

changes the dimensions of dd from 61x50000 to 61x0 because k is 0, which I didn't know was possible in matlab, so that's cool.  

Also, line 32 returns all zeros, but dataRAW looks real.

[row, col, mu] = isolated_peaks_new(dataRAW, ops);

Any insights would be appreciated.

 - Samer Masri

**Comment by tchemaly** [comment]
I have the same problem.

Error using gpuArray/subsref
Subscript indices must either be real positive integers or logicals.

Error in extractTemplatesfromSnippets (line 56)
wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));

Error in extract_spikes (line 42)
[wTEMP, wPCA]    = extractTemplatesfromSnippets(rez, NrankPC);

Error in ksGUI/runSpikesort (line 786)
                [obj.rez, st3, tF]     = extract_spikes(obj.rez);

Error in ksGUI>@(~,~)obj.runSpikesort() (line 337)
                'Callback', @(~,~)obj.runSpikesort());
 
Error while evaluating UIControl Callback.

**Comment by gawygawy** [comment]
i got the same error too! 

Error using gpuArray/subsref
Subscript indices must either be real positive integers or logicals.

Error in extractTemplatesfromSnippets (line 56)
wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));

Error in extract_spikes (line 42)
[wTEMP, wPCA]    = extractTemplatesfromSnippets(rez, NrankPC);

**Comment by bjslater13** [comment]
I am getting the exact same error, just an FYI my data sorted just fine with the older kilosort2.
https://github.com/MouseLand/Kilosort/releases/tag/v2.0

I'm using a linear 64ch probe, same settings for both k2 and k2.5/3. I'll look into the differences and see if I can find anything, but it might be worth just using k2 until there's a solution.

**Comment by marius10p** [comment]
I am not sure these are all the same problem. The error occurs when no spike is found in the raw data. This usually happens due to bad configurations such as wrong numbers of channels, or the data being transposed. Did you all double check in the GUI that the data looks how you'd expect it to? Post some screenshots here please. 

@bjslater13, yours looks like a ks3-specific problem, since it worked with ks2. Have you tried ks2.5? That might give us more information.

**Comment by smasri09** [comment]
Thanks for the comments.  It seems that kilosort 2 does fix that error in my case as well.  The data was always presented properly in time in the gui, so I don't think it was transposed.  I have a new error:
error running kilosort! k must be less than or equal to n.

I don't know what k and n are referring to.  I tried altering ops.reordering and ops.nskip with no luck.

**Comment by marius10p** [comment]
Kilosort2 does not have that spike detection step. You can try reducing ops.spkTh to -4 or something like that so it can catch more spikes for that step. Are your spikes upside-down by any chance, with the voltage going positive? 

To get more informative errors, run Kilosort from a script.

**Comment by smasri09** [comment]
Well after some digging on errors in ks2 that seemed different, I see it's actually the same problem!  This part of my last post.  The data is real, and I've been to sort neurons manually, it's also quite noisy.  I played with threshold, nskip and ordering variables in ops.

On debugging, I'm noticing that 1) dd is a large array of zeros, and 2) this line

% discard empty samples
dd = dd(:, 1:k);

changes the dimensions of dd from 61x50000 to 61x0 because k is 0, which I didn't know was possible in matlab, so that's cool.

**Comment by smasri09** [comment]
To use the scripted version? The output is below.  My suspicion is that kilosort is having these issues because my recordings have noise with much higher amplitude than the spikes.  Could that lead to these issues? This is the error output:


Error using randperm
K must be less than or equal to N.

Error in extractTemplatesfromSnippets (line 58)
wTEMP = dd(:, randperm(size(dd,2), nPCs));

Error in learnTemplates (line 19)
[wTEMP, wPCA]    = extractTemplatesfromSnippets(rez,
NrankPC);

Error in learnAndSolve8b (line 15)
    rez     = learnTemplates(rez, iorder0);

Error in main_kilosort_edit (line 41)
rez = learnAndSolve8b(rez);
 


And here is console output before the error:
>> main_kilosort_edit
Looking for data inside C:\Users\Samer\Desktop\WhiteMatter\logger_5F1211\runnow\ 
Time   0s. Determining good channels.. 
found 121048 threshold crossings in 177.12 seconds of data 
found 0 bad channels 
Time   7s. Computing whitening matrix.. 
Getting channel whitening matrix... 
Channel-whitening matrix computed. 
Time   8s. Loading raw data and applying filters... 
Time  40s. Finished preprocessing 321 batches. 
random seed for clusterSingleBatches: 1
Obtained 7 PC waveforms in 0.52 seconds 
time 0.59, pre clustered 1 / 321 batches 
Batch 13 has fewer than Nfilt spikes.
Batch 37 has fewer than Nfilt spikes.
Batch 66 has fewer than Nfilt spikes.
Batch 67 has fewer than Nfilt spikes.
Batch 68 has fewer than Nfilt spikes.
Batch 72 has fewer than Nfilt spikes.
Batch 73 has fewer than Nfilt spikes.
Batch 81 has fewer than Nfilt spikes.
Batch 85 has fewer than Nfilt spikes.
Batch 86 has fewer than Nfilt spikes.
Batch 96 has fewer than Nfilt spikes.
Batch 100 has fewer than Nfilt spikes.
Batch 101 has fewer than Nfilt spikes.
Batch 106 has fewer than Nfilt spikes.
Batch 107 has fewer than Nfilt spikes.
Batch 112 has fewer than Nfilt spikes.
Batch 113 has fewer than Nfilt spikes.
Batch 114 has fewer than Nfilt spikes.
Batch 115 has fewer than Nfilt spikes.
Batch 116 has fewer than Nfilt spikes.
Batch 119 has fewer than Nfilt spikes.
Batch 120 has fewer than Nfilt spikes.
Batch 121 has fewer than Nfilt spikes.
Batch 124 has fewer than Nfilt spikes.
Batch 198 has fewer than Nfilt spikes.
Batch 199 has fewer than Nfilt spikes.
Batch 200 has fewer than Nfilt spikes.
Batch 201 has fewer than Nfilt spikes.
Batch 207 has fewer than Nfilt spikes.
Batch 210 has fewer than Nfilt spikes.
Batch 297 has fewer than Nfilt spikes.
Batch 305 has fewer than Nfilt spikes.
Batch 310 has fewer than Nfilt spikes.
Batch 311 has fewer than Nfilt spikes.
Batch 312 has fewer than Nfilt spikes.
Batch 314 has fewer than Nfilt spikes.
Batch 315 has fewer than Nfilt spikes.
Batch 316 has fewer than Nfilt spikes.
Batch 317 has fewer than Nfilt spikes.
Batch 318 has fewer than Nfilt spikes.
Batch 319 has fewer than Nfilt spikes.
Batch 320 has fewer than Nfilt spikes.
Batch 321 has fewer than Nfilt spikes.
time 0.04, compared 1 / 321 batches 
time 8.78, Re-ordered 321 batches.

**Comment by putnampt** [comment]
I am also receiving this error message.

>  Error using gpuArray/subsref
> Subscript indices must either be real positive integers or logicals.
> 
> Error in extractTemplatesfromSnippets (line 56)
> wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));
> 
> Error in extract_spikes (line 42)
> [wTEMP, wPCA]    = extractTemplatesfromSnippets(rez, NrankPC);
> 
> Error in ksGUI/runSpikesort (line 795)
>                 [obj.rez, st3, tF]     = extract_spikes(obj.rez);
> 
> Error in ksGUI>@(~,~)obj.runSpikesort() (line 339)
>                 'Callback', @(~,~)obj.runSpikesort());
>  
> Error while evaluating UIControl Callback.

I am working with data from a 16 channel probe that has been written into a binary int16 .dat file. The GUI seems to be handling the data at some level; preprocessing works well and the visual traces of the analog clearly show spikes.

Here is an image of what the GUI looks like:
![image](https://user-images.githubusercontent.com/5158403/112670560-714ab900-8e37-11eb-91a9-766029628dbe.png)

This error occurs when trying to 'Spikesort'. I have verified that there are spikes on all channels, and that the data seems to be properly formatted in the .dat file. I have tried variety of thresholds, both negative and positive and have receive the same error.

**Comment by nicolomeneghetti** [comment]
Dear PutNampt,

I have solved the same issue by simply changing the channel map configuration. 
Specifically, by using the [new] option of 'Select Probe Layout' of the GUI I have created a custom made channel map of my 16 channels linear probe with 50um of inter electrode spacing as so (see attached screenshot):
![Immagine_screee_channel_map_kilosortGUI](https://user-images.githubusercontent.com/82510633/114835731-2a1a6d00-9dd2-11eb-9d16-c7879bdcf620.png)

**Comment by gaito** [comment]
Hello, I was just commenting to see if anyone got the aforementioned issue resolved? I am getting it as well!

> Error using gpuArray/subsref
> Subscript indices must either be real positive integers or logicals.
> 
> Error in extractTemplatesfromSnippets (line 56)
> wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));
> 
> Error in extract_spikes (line 42)
> [wTEMP, wPCA] = extractTemplatesfromSnippets(rez, NrankPC);
> 
> Error in ksGUI/runSpikesort (line 786)
> [obj.rez, st3, tF] = extract_spikes(obj.rez);
> 
> Error in ksGUI>@(~,~)obj.runSpikesort() (line 337)
> 'Callback', @(~,~)obj.runSpikesort());
> 
> Error while evaluating UIControl Callback.

**Comment by bjslater13** [comment]
My issue had to do with the path referencing the incorrect function with
the same name.

**Comment by gaito** [comment]
Which function was it? @bjslater13

**Comment by bjslater13** [comment]
I had a previous installation of kilosort it was referencing. I can't
recall the exact function. I ended up resetting the path and adding things
one by one.

**Comment by nikolaskaralis** [comment]
In case anybody has this issue again in the future and is reading this thread:
There are few different issues discussed above.

For the issue with the error:
"Error using randperm
K must be less than or equal to N."

The real problem is not having few spikes, so it cannot really be addressed by changing the parameters.
The core issue that leads to this error is improper assignment of xcoords and ycoords.

Ensure that your channels have distinct ycoords and it should fix this problem.

**Comment by Feikai-Lin** [comment]
Does anyone have any update on this? 

I also got this problem when I used simulated data generated from main_eMouse_drift.m in the Kilosort3_main folder (with all default configuration/parameters except removing ref channel in make_eMouseChannelMap_3B_short.m). I have checked in GUI that the simulated data was generated and imported properly (with descent amount of spikes etc). 

Here are the error messages I got: 
Error using gpuArray/subsref
Subscript indices must either be real positive integers or logicals.

Error extractTemplatesfromSnippets (line 56)
wTEMP = dd(:, round(linspace(1, size(dd,2), nPCs)));

Error datashift2 (line 4)
[wTEMP, wPCA]    = extractTemplatesfromSnippets(rez, NrankPC);

Error main_eMouse_drift (line 72)
        rez = datashift2(rez, 1); % last input is for shifting data

While debugging, I find that in extractTemplatesfromSnippets.m: 
(line 14) fid = fopen(ops.fproc, 'r'); -------> giving that fid = 3
(line 22) dat = fread(fid, [ops.Nchan NT], '*int16'); -------> dat is a large array or zeros.  
Since temp_wh.dat is actually quite large (4GB), I am not sure if problems occur in fread or the temp_wh.dat; 

I only have ks3 in our workstation (along with Matlab2018b with cuda9.1 on win10). Also I noticed that #320 main_eMouse_drift.m in ks3 hasn't been updated yet. However, the errors I got appeared before learnAndSolve8b. 

Any suggestions would be appreciated.
